open_lm/data.py

```diff
@@ -577,3 +578,87 @@ def get_data(args, epoch=0, tokenizer=None, skip_train=False, floor=True):
             data["val_list"].append(data_val)
 
     return data
+
+
+# preprocessing tokens and sampling chunks
+
+
+def squash_tok(seqs, mask, pad):
+    # Get the count of non-delimiter elements in each batch
+    count_non_squash = mask.sum(dim=1)
+
+    # Create a tensor of pad tokens with the same shape as the input tensor
+    out_tensor = torch.full(seqs.shape, pad, dtype=seqs.dtype, device=seqs.device)
+
+    # Place all non-delimiter elements at the left of the output tensor, pad tokens to the right
+    for i in range(seqs.shape[0]):
+        if count_non_squash[i].item() != 0:
+            out_tensor[i, : count_non_squash[i]] = seqs[i, mask[i]]
+        else:
+            out_tensor[i] = seqs[i]
+
+    return out_tensor
+
+
+def mask_sequence(chunk, start_idx, args, ignore_tok=-100):
+    # NOTE: this implementation supports 0 or 1 instance of tok in a sequence.
+    #       if more than one instance appears, the last instace of tok is used.
+
+    tensor = torch.clone(chunk)
+
+    inputs = tensor[:, start_idx : start_idx + args.seq_len]
+    targets = tensor[:, start_idx + 1 : start_idx + args.seq_len + 1]
+
+    if args.target_mask_left is not None:
+        tok_positions_targets = targets == args.target_mask_left
+
+        # construct cumulative mask for positions before (last) tok (if it appears)
+        cumsum_mask = tok_positions_targets.flip(dims=[-1]).cumsum(dim=-1).flip(dims=[-1])
+
+        # create mask for positions before (last) tok in each row (batch)
+        tok_mask = cumsum_mask > 0
+
+        if args.squash_mask_left:
+            # retain tok in the output as makes squashing easier
+            tok_mask &= ~tok_positions_targets
+
+            # ensure we dont modify input
+            targets = torch.clone(targets)
+            targets[tok_mask] = ignore_tok
```

Comment:
I'm struggling a bit in this workflow. Can you add an example like the following in comments? I think it would really help. I attempted below, but I'm getting something wrong in the logic.

```# Let
# L = args.target_mask_left
# _ = ignore_tok
# P = pad token (args.target_mask_individual)
inputs                = [a, b, c, L, d, e, L, f]
targets               = [b, c, L, d, e, L, f, g]

tok_positions_targets = [0, 0, 1, 0, 0, 1, 0, 0]
cumsum_mask           = [2, 2, 2, 1, 1, 1, 0, 0]
tok_mask              = [1, 1, 1, 1, 1, 1, 0, 0]
# after tok_mask &= ~tok_positions_targets
tok_mask              = [1, 1, 0, 1, 1, 0, 0, 0]
# after targets[tok_mask] = ignore_tok; let _ = ignore_tok
targets               = [_, _, L, _, _, L, e, f]
selected_inputs       = [1, 1, 1, 0, 1, 1, 0, 1]
selected_targets      = [1, 1, 0, 1, 1, 0, 1, 1]
# after squash, let P = pad token (args.target_mask_individual)
inputs                = [a, b, c, d, e, f, P, P]
targets               = [_, _, _, _, e, f, P, P]
```

Reply:
this is a great idea! admittedly the logic is kinda confusing, tried to correct the hand-simulation in place
```
# L = args.target_mask_left
# _ = ignore_tok
# P = pad token (args.target_mask_individual)
inputs                = [a, b, c, L, d, e, L, f]
targets               = [b, c, L, d, e, L, f, g]

tok_positions_targets = [0, 0, 1, 0, 0, 1, 0, 0]
cumsum_mask           = [2, 2, 2, 1, 1, 1, 0, 0]
tok_mask              = [1, 1, 1, 1, 1, 1, 0, 0]
# after tok_mask &= ~tok_positions_targets
tok_mask              = [1, 1, 0, 1, 1, 0, 0, 0]
# after targets[tok_mask] = ignore_tok; let _ = ignore_tok
targets               = [_, _, L, _, _, L, f, g] # LAST 2 ELEMENTS HERE ARE f, g, NOT e, f
selected_inputs       = [1, 1, 1, 0, 1, 1, 0, 1]
selected_targets      = [1, 1, 0, 1, 1, 0, 1, 1]
# after squash, let P = pad token (args.target_mask_individual)
inputs                = [a, b, c, d, e, f, P, P]
targets               = [_, _, _, _, f, g, P, P] # PROPOGATING HERE
```

Reply:
ahh perfect, thanks! yeah if you can add this in as comments it would help. You don't need all the steps (since the comment may go out of date when we change the code), maybe just the original inputs/targets and final inputs/targets would help